### ISSUE TITLE
Always throw from SwitchToMainThreadAsync if the token is canceled

### ISSUE THREAD

**Issue by CyrusNajmabadi** [microsoft/vs-threading#434]
For more context, see thread here: https://github.com/dotnet/roslyn/pull/31787

Specifically starting here: https://github.com/dotnet/roslyn/pull/31787#issuecomment-447502124

TLDR: Roslyn switched some explicit usages of TPL+SyncContexts to use JTF for switching to the UI thread.  This ended up causing crashes due to Roslyn depending on TPL behavior that JTF doesn't provide.  Specifically, roslyn often 'chains' tasks along.  For example, it might have the following sort of chain:

```
Task1(ui thread) -> ContinuesWith -> Task2 (expensive BG code) -> ContinuesWith -> Task3 (ui thread)
```

This is common for us so that we can update shared state *and* often let teh rest of VS know about something important in a STA manner.

While these tasks are running we are still processing inputs and, on the UI thread, we may decide to cancel this chain of work.  For example, while Task2 is executing, we may end up cancelling on hte UI thread because of something the user did.  With TPL, we had behavior whereby even if TPL decided Task3 was to run (because, say, Task2 had completed and things had been scheduled), once Task3 actually executed, it would always see the cancellation made on the UI thread.  This happened automatically by TPL.  Before starting to run Task3, it would do a final cancellation check and would then throw in this case.  In essence, we depended on the TPL behavior that any changes made on one thread would be seen by later tasks that would run on that thread.  Since we canceled on hte UI thread, it was certain that any other tasks intended ot run on the UI thread would always see that cancellation.

JTF does not have this behavior.  Even if we cancel on the UI thread, JTF will allow a descendant task to both switch to the UI thread and continue running.  This broke the expected behavior we god from TPL.

Fortunately, in the linked issue, this caused a crash, helping to track this down.  But, far more often, when roslyn has encountered some sort of race like this, it has led to data corruption which often just leads to broken behavior for the user.  Usually because data is updated that should not be updated.

--

The ask here is for JTF to have an option on these methods to behave like TPL.  Specifically, one should be able to ask it to switch to the UI thread, and have it check+throw once it gets to the UI thread if previous code on teh UI thread canceled that exception. 

While i would like this to be the default, it is recognized that this could be considered a fairly drastic change in behavior that would negatively affect others.

**Comment by AArnott** [comment]
(I wrote this up for your original PR, but now that we have an issue dedicated to this discussion, I moved it here)

Good points, @CyrusNajmabadi.

Not to take away from anything you said, since from where you're coming from it all sounds perfectly reasonable, but I tend to go a different way. 

TPL is largely dead. I say that because the general consensus is that folks should typically *not* use `Task.Factory.StartNew` or `Task.ContinueWith` any more, but rather use async/await patterns. There are a lot of reasons for this, and I'll assume you're familiar with them. As such, JTF was very specifically written to appeal to the async/await coding patterns *rather* than TPL. You'll notice we didn't even write a JTF-aware `TaskScheduler` so that folks who use TPL patterns could avoid deadlocks. That's partly because TPL `TaskScheduler` is so archaic that it doesn't actually support async tasks. I'm not knocking technology here, to be clear. It was written before async/await, and some of async/await is even built on TaskScheduler. But JTF SwitchToMainThreadAsync is a method to be called by *async* methods -- not by delegates queued with TPL. So they really are two very different worlds (async/await vs. TPL scheduled tasks). And the rules don't have to be the same. If we modeled JTF after 4.0-era TPL instead of after async/await, we'd be designing for the rocket scientist crowd (e.g. @CyrusNajmabadi) instead of for the average engineer, who has to work just to grasp how async/await work.

So if we are targeting folks who first learned async/await and never knew about .NET 4.0-era scheduled Tasks, then when TPL precedents and async/await precedents conflict, we should model STMTA the way other async methods tend to work to fit the expectations of the typical consumer. And in the async/await world, I don't think I've ever seen an async method that was wise to call with a `CancellationToken` and assume that the method's successful completion means the token wasn't canceled. So such a guarantee on JTF's STMTA didn't seem to have an interested audience (till now, at least).

I acknowledge that for the Roslyn engineering team that following TPL rules would make your migration easier given your background, but with respect, you're not the average customer. :wink:

**Comment by sharwell** [comment]
> The ask here is for JTF to have an option on these methods to behave like TPL. Specifically, one should be able to ask it to switch to the UI thread, and have it check+throw once it gets to the UI thread if previous code on teh UI thread canceled that exception.

I do not believe this is a correct representation of the way the TPL behaves. More specifically, the TPL does not provide any equivalent to `SwitchToMainThreadAsync`, which means transitioning to JTF puts us in a place where the code is easier to read but there is no direct behavior mapping.

**Comment by AArnott** [comment]
So exploring your API request, @CyrusNajmabadi, there are two possibilities that occur to me (don't sweat the names I'm using... it's the shape I mean to discuss at this stage):

1. Add a new method: `JTF.SwitchToMainThreadThrowOnCancellationAsync(CancellationToken token)`
1. Add an overload to our existing one: `JTF.SwitchToMainThreadAsync(token, alwaysThrowOnCancellation: true)`

Either way, it requires some explicit opt-in. For the change in behavior to be clear, the naming has to be fairly verbose. And someone who has never seen it before has to study and think about it before realizing the nuanced impact it has. The "normal" behavior (given the original method/overload) still looks correct to the casual code reviewer, so where an opt-in to the new behavior is important, it could easily be missed.

Still a 3rd possibility is that you have a specifically configured `JoinableTaskFactory` *instance*. We already have a precedent for `JoinableTaskFactory` derived types that have specific priorities for scheduling work. We could allow a JTF instance to influence the behavior you're looking for. That would allow Roslyn to automatically pick up the desired behavior everywhere -- provided you're using the right JTF instance. So that's good in that it's easy to do the right thing (for your code base). But it *may* be harder to audit when you're not sure which JTF instance you're calling, or even aware that you're supposed to be calling a particular one.

I'm curious which one of these appeal to you folks.

I'm not a fan of the first two options especially, since today we already have a STMTA+ThrowOnCancellationRequested pattern that folks who have your requirement can already do. Adding another way, that is almost as tedious to code up (we're just translating a method call into a parameter or method name), wouldn't pass the .NET Framework API cost test, IMO.

**Comment by sharwell** [comment]
I'm less interested in the pattern we have now, and more interested in the subset case where we avoid executing the continuation on the main thread _if cancellation occurs on the main thread_.

**Comment by AArnott** [comment]
> more interested in the subset case where we avoid executing the continuation on the main thread if cancellation occurs on the main thread.

Can you explain this more? I'm not sure what you're arguing for. The subset you describe, from what I've heard on your PR and by email, is what Roslyn's case is in the first place. 

> If cancellation occurs on a different thread, we have a race condition anyway and the cancellation check would only provide a false sense of security.

See, I've never seen cancellation as a way to protect data. Even the naming around a cancellation token ("is cancellation *requested*") indicates to me that it's merely a request, and that it may or may not be fulfilled, sooner or later. IMO if I had data to protect, I would be very explicit about it, and consider anything else (like cancellation handling) to be an optimization on top of it.

**Comment by CyrusNajmabadi** [comment]
> I do not believe this is a correct representation of the way the TPL behaves. More specifically, the TPL does not provide any equivalent to SwitchToMainThreadAsync

isn't that equivalent to scheduling a task on the appropriate sync context?

**Comment by CyrusNajmabadi** [comment]
>See, I've never seen cancellation as a way to protect data. Even the naming around a cancellation token ("is cancellation requested") indicates to me that it's merely a request, and that it may or may not be fulfilled,

It may not be fulfilled *in that* you might write code to not respect it.  However, we've written code to respect it.  Note: none of this was strange when we did any of our discussions with the TPL people about this.  Why would we need anything else when Cancellation already behaved this way?

i.e. we could have had *another* mechanism.  one where we an antecedent task set some bit, and a later task checked that bit.  But we'd also want that later task to throw... so that we wouldn't have to check at every layer.  And we'd then have to catch and filter out that particular excpetion.

Except... now we've reinvented cancellation tokens :)

> IMO if I had data to protect, I would be very explicit about it, and consider anything else (like cancellation handling) to be an optimization on top of it.

But, to me, we are being explicit.  I am saying "all future work is canceled".  We've designed our systems such that data mutation can only happen on tasks that will, **by design**, see that notification.  So we've been explicit, and those descendent tasks must see that request because they are going to run on the same thread where we made that statement.

It's a 'request' in that there's no way to 'force' that thread (other than by aborting) to listen.  But that would be the same with any other mechanism we made.  If we set another boolean somewhere, that would *still* be a request, because we'd still depend on that other thread respecting the boolean.

It's all ultimately cooperative.  Except that we leveraged the existing mechanism that was already well baked into TPL instead of going NIH and building our own :D

**Comment by AArnott** [comment]
> isn't that equivalent to scheduling a task on the appropriate sync context?

Yes, roughly. But that's only equivalent at the highest level. Beneath that conceptual equivalence, almost all the mechanics of how it works underneath are different. And it looks completely different even at the surface when you consume the API. So aside from "they both get you to the UI thread", they're actually quite different.

**Comment by CyrusNajmabadi** [comment]
> Either way, it requires some explicit opt-in. For the change in behavior to be clear, the naming has to be fairly verbose. And someone who has never seen it before has to study and think about it before realizing the nuanced impact it has. The "normal" behavior (given the original method/overload) still looks correct to the casual code reviewer, so where an opt-in to the new behavior is important, it could easily be missed.

Importantly, we can add a threading analyzer that says that we cannot call that other API as it doesn't match teh semantics we except.

**Comment by CyrusNajmabadi** [comment]
>  they're actually quite different.

This is not a compelling argument for me.  If htey are 'completely different', then i would argue that Roslyn should not use JTF.   We should not be mixing/matching totally different threading/concurrency systems.  Because that's exactly how we end up in the state we got into, where we have broken behavior and not a single person (with probably 50+ years of combined experience around threading/TPL/JTF) had a clue that there was a problem. 

You're telling me they're different.  I get that.   I'm saying: i don't want to mix/match such differing systems, because it just adds complexity in a system we have spent umpteen years trying to keep simple in order to prevent these problesm in the first place.  So, given that, my end feelings are:

1. Roslyn should not use TPL.  (massively expensive, likely not going to happen any time soon).
2. Roslyn should not use JTF.  Unfortunate.  JTF has nice benefits.  I would like to use it.  
3. JTF makes concessions to behave more like TPL.  My preference.  Makes it feel like something we can more safely move to.  Makes JTF and TPL work better together.  Makes it easier to write analyzers that check for problems and which guides the team to use the right stuff.  etc. etc.

**Comment by AArnott** [comment]
> So we've been explicit, and those descendent tasks must see that request because they are going to run on the same thread where we made that statement.

That's fine. I'm not against you designing your system as you have. If you want to constrain yourself to one thread, and explicitly cancel and check tokens, that's great. Nothing wrong with reusing CancellationTokens for this -- don't reinvent them. 

My argument is that just because TPL behaved a particular way, JTF doesn't necessarily have to conform to that precedent (that was never a precedent for JTF to begin with). And yet you can continue being explicit with cancellation -- but as you migrate from TPL to async/await, you need to change your assumptions about how cancellation tokens are honored from "TPL will handle it for us" to "we need to throw if it's canceled".

**Comment by CyrusNajmabadi** [comment]
Moved ths from the other thread:

>    TPL is largely dead.

This is the first i'm hearing anything of the sort... can you link to any sort of statements to that effect?

>    I say that because the general consensus is that folks should typically not use Task.Factory.StartNew or Task.ContinueWith any more, but rather use async/await patterns

Sure... but what i was talking about still applies with async/await. I was simply talking aobut .ContinueWith since htat's hte underlying mechanisms this boils down to. It doesn't actually depend on anything async/awaity.

i.e. we use async/await heavily in Roslyn. But we also expect that code that is scheduled back to the UI thread (i.e. because we might have used .ConfigureAwait(true) on code that started from the UI thread) will still see cancellations made on the UI thread.

>    And the rules don't have to be the same.

I do agree on that. But that means that from the Roslyn side of things, i'd prefer to not mix and match. Because, well, this is the very realistic outcome. Concurrency is already extremely hard. And one of the ways that Roslyn has managed that was to be very very very very very careful about things. We intentionally do not try to be fancy most of the time. We intentionally have tried to only use a small part of the surface area that TPL provides. We intentionally have opted for simplicity, consistency, and clear semantics above everything, so that we can deliver a correct system first, with everything else following.

Indeed, for Roslyn IDE, the core goals are:

Correctness first. It needs to do things properly, or else, there's no point.
Non-blocking second. Given 'correctness', our primary use for TPL is simply to get things off the UI thread.
Performance. We can benefit here by also being able to leverage threadpools. But that's really just an added benefit in some cases.

Many of us cut our teeth on the preceding systems that were overly complex and basically had race conditions galore. We intentionally moved far away from that for Roslyn, realizing that even having concurrency/asynchrony in the first place was already increasing complexity by orders of magnitude. Given that, we intentionally tried to keep things as simple as we possibly could. And that means carving out a narrow section of TPL we could completely understand (and teach to the team) and religiously following patterns and practices to make it possible to sanely develop and maintain this code without having race conditions every day of hte week. It's worth pointing out that some parts of Roslyn didn't go this route, and we're still paying the price with flakeyness there to this day.

>    but with respect, you're not the average customer.

We're not. But, tbh, i think we're an important customer in that we strive for simplicity, safety, consistency and clarity much of the time. IMO, a lot of what we've learned, and a lot of hte patterns and practices we've taken to allow for concurrency, while still be safe/correct, would be valuable for others to follow. These are tactics created to actually reign in the complexity of concurrency, and (as we can see here) when the system gets more complex and inconsistent, even incredibly capable and experienced teams can introduce very worrying bugs.

**Comment by CyrusNajmabadi** [comment]
> My argument is that just because TPL behaved a particular way, JTF doesn't necessarily have to conform to that precedent (that was never a precedent for JTF to begin with). And yet you can continue being explicit with cancellation -- but as you migrate from TPL to async/await, you need to change your assumptions about how cancellation tokens are honored from "TPL will handle it for us" to "we need to throw if it's canceled".

That is certainly an option.  But i personally don't think it's a good one.  It violates a core engineering goal Roslyn had around concurrency which was to KISS.  

Needing to understand these sort of subtle difference is the antithesis of KISS.  Note: if JTF expects that it is used in isolation of TPL, then it's *certainly* the case that it is abiding by KISS *itself*.  The problem then exists for the team that wants to have a reasonable migration path from TPL to KISS..

So, my question for you is: would you be amenable to making changes that go against JTF's internal simplicity, in order to help teams with the transition path from TPL to it?  I personally think it would be a good idea.  :)

**Comment by sharwell** [comment]
@CyrusNajmabadi It's not vs-threading that's violating KISS here. Roslyn using application of CancellationToken as a basis for algorithm correctness is already too complex; vs-threading operates under even simpler considerations.

**Comment by AArnott** [comment]
>> TPL is largely dead.

> This is the first i'm hearing anything of the sort... can you link to any sort of statements to that effect?

Maybe "dead" is too strong a word. But Stephen Toub I believe agrees, and other influential folks (including David Fowler) have repeatedly said that with async/await, the .NET 4.0 era TaskScheduler patterns are now in the domain of "rocket scientists". This is because it's tedious, hard to code right, and even harder to code more efficiently than async/await is. So much work went into the compilers to minimize allocations and overhead for async/await, that it's become quite difficult to write TPL code that is more efficient (GC pressure-wise, at least) than async/await.

**Comment by CyrusNajmabadi** [comment]
> Maybe "dead" is too strong a word

![image](https://user-images.githubusercontent.com/4564579/50112658-9e3da980-01f4-11e9-8da7-0c74b5e17be0.png)

:D

**Comment by CyrusNajmabadi** [comment]
Look.  I'm not disagreeing with your guys' points.   :)

I'm simply stating: this issue is a clear indication of something concerning.  I'm coming from the direction of caring about the engineering here and the concerns about reliability and correctness.  It's not an exaggeration to say that issues here are some of the worst Roslyn has ever had to deal with. 

As such, i come from the perspective of wanting as safe and understandable a system as possible.  And when 'the new api' behaves differently from 'the api that was the one to use up till now', and when we have the *direct evidence* of the sort of problem you run into becuase of that, then all the theoretical hopes and desires around simplicity and whatnot get thrown out the window for me :)

At this point, i can't really add more.  The request has been made for a API that behaves more like TPL.  If you want to add it, great.  If you don't that's fine.  

--

In the meantime, if there are other areas where you know that JTF deviates from TPL-like behaviors, can you let me know?  I want to know if there are things i need to be carefully considering when i review code that uses JTF.  Thanks! :)

**Comment by sharwell** [comment]
At this point I lean against API or behavior changes. The one motivating example presented so far is a case of desiring that implicit cancellation be respected at a location simply to avoid calling `ThrowIfCancellationRequested`. It's generally safe to omit the latter without impacting correctness, and in the remaining cases it seems desirable for the cancellation to be explicit.

**Comment by AArnott** [comment]
> Specifically, TPL has the bahvior that any .ContinueWith'd task will see the writes of its antecedent tasks. This is true regardless of waht threads either of those tasks run on.
> Again, it wasn't necessary that TPL ensure that. It would have been simpler for it not to.

I can't imagine a world where TPL didn't provide that guarantee. `Task.ContinueWith` by definition of the very method name suggests that the next task won't execute before the prior one completes. Why do you say TPL could have gotten away without that guarantee?

> So, when i say "KISS" i'm not saying from the library's perspective. I'm saying "from the consumers" perspective.

To be clear, JTF is anything but simple internally. I am fully in favor of designing for simplicity in consumption of the library. But that's why I hesitate to add more API -- not because JTF becomes more complicated to implement but because it becomes more complicated to consume.

>  It's much simpler to me that if i cancel a token on a particular thread that i can know that dependent tasks will certainly not execute on that same thread. 

Your scenario isn't a dependent task. In TPL there were clear task->continuewith dependencies where one task depended on another. But in async/await, there aren't explicit task dependencies. JTF had to *invent* that concept. In fact, in your case it's not even an implicit dependency, AFAIK. The use case Roslyn is looking at is rather a case of an _unrelated_ async method requesting the main thread, after which someone *else* cancels the token. There's no dependency between two tasks there.

**Comment by CyrusNajmabadi** [comment]
> So your option 2 (use TPL and not JTF) is simply not an option that I can imagine a working system for. 

That's obviously untrue.  Roslyn operated for like 10+ years with TPL and without JTF, while being able to use the UI thread.  We designed for this and we very carefully and conscientiously implemented many components that worked fine in this regard.

And, IMO, this was a *better* state to be in.  We never took it for granted that we could use the UI thread.  Instead, we understood very well the risks involved and we designed and implemented our features very carefully to be explicit about how this works so it would be all safe.  This ended up with very understandable systems IMO.  Because we couldn't just easily do this sort of thing, we had to be very careful and clear about how we actually did it.  And, in the end that was usually much better, because we weren't haphazardly doing things (like causing UI calls) at times when it was inappropriate.

So again, i reject the idea that you cannot have "a working system for" this case.  We accomplished it quite well in a huge number of different domains. i.e. the number of different sorts of UI services that roslyn needs to interact with (while also wanting to be async+bg oriented) is huge.  But we were able to conscientiously design ways to deal with this in all circumstances.  And, as stated, i think forcing the separation was a *good* thing.

--

Also, since moving to JTF, i've had to block a bunch of PRs that have tried to now use JTF at inappropriate times.  In the past, even trying to use the UI thread in the way the PRs were written simply wouldn't work.  You would deadlock, and you'd immediately know this was a bad thing.  With JTF, there is the sirens calls of "it's ok to make this call to the UI thread, and JTF will take care of it for you", when really it isn't.  Yes, you won't deadlock.  But you'll potentially massively degrade the experience of that feature.

Sometimes it's better to actually consider the UI thread something you should be extremely wary of using, and something that if you misuse causes major problems.  It helps force you away from it, and it helps force you to design systems that are much more conscientious about how it is used.  That's the approach Roslyn took for many years, and i think it very much was a good thing.

**Comment by CyrusNajmabadi** [comment]
> I can't imagine a world where TPL didn't provide that guarantee. Task.ContinueWith by definition of the very method name suggests that the next task won't execute before the prior one completes. Why do you say TPL could have gotten away without that guarantee?

I can't imagine a world where JTF didn't provide the guarantee we're speaking about now.  And yet here we are :)

**Comment by CyrusNajmabadi** [comment]
> But in async/await, there aren't explicit task dependencies.

Whether it is implicit or explicit doesn't make a difference to me.  The same behavior should hold for implicit dependencies IMO.  Having things be different just makes things *complex* on the *consumption* side.

**Comment by AArnott** [comment]
> I can't imagine a world where JTF didn't provide the guarantee we're speaking about now. And yet here we are :)

I thought you were going to say that. But the difference is, `ContinueWith` screams "this happens later" within its own domain (and more broadly, I argue). But within the domain of async methods, a cancellation token has *no guarantee whatsoever* of being honored. That's a very strong precedent that JTF didn't set -- we just live with it.

**Comment by CyrusNajmabadi** [comment]
> ContinueWith screams "this happens later" within its own domain (and more broadly, I argue). 

Yes.... and when i say "ContinueWith" with this cancellation token, and it says "this will happen later", and i then cancel the token, i *expect* that the cancellation token would be respected if that later point is on my same thread...   :)

**Comment by AArnott** [comment]
> > So your option 2 (use TPL and not JTF) is simply not an option that I can imagine a working system for.

> That's obviously untrue. Roslyn operated for like 10+ years with TPL and without JTF, while being able to use the UI thread. We designed for this and we very carefully and conscientiously implemented many components that worked fine in this regard.

And yet, Roslyn eventually switched to JTF. Why is that? Did they need something that they couldn't get otherwise? As scenarios expand, and the need to interop better with other VS components grows, JTF becomes a more pressing requirement for anything that is async and may require the UI thread, such as Roslyn.

**Comment by CyrusNajmabadi** [comment]
> And yet, Roslyn eventually switched to JTF. Why is that? 

Don't know.  I was very resistant to it.  But i'm also off the team.  I'm not apriori against it being used.  But my preference is toward simplicity for *precisely* the reasons that started this entire conversation.

> As scenarios expand, and the need to interop better with other VS components grows, JTF becomes a more pressing requirement for anything that is async and may require the UI thread, such as Roslyn.

Honestly, i don't see why that's the case at all.  Roslyn interops with dozens (maybe hundreds) of VS components.  And due to the intense UI-affinitized behavior of VS, haivng to be UI thread-aware was baked into Roslyn-IDE from day one.  

In 100% of cases, we were able to properly do things in a conscientious and methodical fashion.  

--

Note: again, i'm not against JTF.  What i'm pushing for is a path to use JTF that has few "gotchas" as possible. I'm surprised that this would even be contentious.  We're all engineers here.  We know how problematic it can be when systems behave in subtly different manners.   I'm just trying to find effective ways to be able to be an engineer that can use these libraries in a safe manner to deliver an altogether great product to customers.

Also, as an aside, i'm going to ignore me and just talk about Sam/Jason.  They are, without question, two of the absolute best people i've know when it comes to understand the deep complexities around threading/async/await/tasks/VS/STA/ui-threads/component-designs/etc.  Even with both of them carefully examining the changes to JTF (and even with me closely examining each line and giving lots of feedback on these PRs) this issue was missed.  That's a big red flag that there's an issue here.  I'd like Roslyn to be able to use JTF effectively.  But part of that, to me, is having confidence that the rules and behaviors i've been able to trust from the TPL world are held in the JTF world.

So, my hope would be that thsi is the only place they deviate.  But i have no confidence or reason to believe that's the case. That's why i asked if you could enumerate any other known differences.  I don't want to continue learning about the differences because of ccrashes (or worse yet, subtle corruptions that may take months to discover). I'd like to know about them up front so i can be auditing code, and keeping that all in mind when i do PRs :)

**Comment by AArnott** [comment]
> Sometimes it's better to actually consider the UI thread something you should be extremely wary of using, and something that if you misuse causes major problems. It helps force you away from it, and it helps force you to design systems that are much more conscientious about how it is used. That's the approach Roslyn took for many years, and i think it very much was a good thing.

:smile: It has come up a few times recently that JTF has made accessing the UI thread *so* easy that perhaps it's a disservice to itself. I always chuckle a bit when we discuss such a sentiment. I'm quite happy with the fact that we've now moved the problem space from "it's too hard to do it" to "it's so easy but we need to think about when to use it". As a product (or actually, a set of products since JTF is now used in several) we've been able to accomplish many things as far as solving UI delays and enabling greater concurrency and thus customer performance wins as a result of JTF making async/await compatible with UI thread requirements and synchronously blocking that UI thread. 

We've heard two engineers' votes here from Roslyn: one for, one against. I'll keep the issue active for a while longer to see if anyone else has votes or arguments for the proposed change.

**Comment by CyrusNajmabadi** [comment]
>We've heard two engineers' votes here from Roslyn: one for, one against. I'll keep the issue active for a while longer to see if anyone else has votes or arguments for the proposed change.

Sounds good :)

**Comment by AArnott** [comment]
Great. So not to continue the debate, but to answer a question you asked that I think leads to additional value we can glean from this discussion...

@CyrusNajmabadi said:

> So, my hope would be that thsi is the only place they deviate. But i have no confidence or reason to believe that's the case. That's why i asked if you could enumerate any other known differences.

Enumerating differences is tough since the two systems are modeled so differently. But I can share common gotchas that I've seen from the many teams that have migrated so far. And we have analyzers that codify guards for those gotchas wherever we have the means (and time) to do so. I had written up a list of a few gotchas, but then realized that really, our [Threading Cookbook and threading analyzers](https://aka.ms/vsthreading) pretty much list all of them (including this one, now that my doc fix PR is in).

**Comment by CyrusNajmabadi** [comment]
This is a *lot* to digest.  I'll try to budget a few hours to try to internalize it all.  It def makes me more wary about hte usage of jtf here.

**Comment by jasonmalinowski** [comment]
@AArnott do we have concrete examples in VS today that rely on the existing behavior for correctness? Between seeing the places where Roslyn is being broken, and the places in the VS code that at least (to me) also look suspicious. I can understand if we want to add an option or analyzer to avoid a breaking change because yes, it's potentially breaking and I don't dispute that. But how much code is relying on this, intentionally? Counting up the uses I'm seeing a bunch that expect/probably want the behavior where cancellation is raised on the UI thread, and I'm not seeing the other; if that's just me looking in the wrong place I'd want to be corrected ASAP.

I do see the "things shouldn't cancel if they've done the work" argument, but I guess that doesn't seem particularly meaningful to me here. I'd consider that a weak guideline at _best_, and we've had good reasons to break that in the past. For example, Roslyn's implementation of cancellation-aware AsyncLazy will return a cancelled task for GetValueAsync(), even if we already had a cached result. We found it was better to give a cancelled task back when somebody wants a syntax tree versus let them start to analyze it, observe the cancellation later, and then give up anyways some time later. Put another way, if we didn't have that in AsyncLazy, we'd probably expect all consumers to do that to the point that we'd require an analyzer, at which point we didn't actually gain anything by following this "rule".

> And yet, Roslyn eventually switched to JTF. Why is that?

I don't consider Roslyn as having moved to JTF, at least not in a meaningful way. Yes, it now consumes the NuGet packages, but the majority of our use is either SwitchToMainThreadAsync, or some limited cases in the VS codebase when we're interacting with legacy components. But we're not writing idiomatic JTF code by any stretch of the imagination, and doing so in many cases is still a API back-compat problem. To me the primary benefit up to this point has just been deleting our code that tried to give us an implementation "am I on the UI thread? I'd like to assert that" and "give me a TaskScheduler for running on the UI thread". If those helpers were available directly there's a lot of places we might have just used those, and I'd still use those. :smile:

> It has come up a few times recently that JTF has made accessing the UI thread so easy that perhaps it's a disservice to itself. I always chuckle a bit when we discuss such a sentiment.

I'd say the don't chuckle; this to me is the truth and why I've loved that Roslyn didn't use JTF. It means when Roslyn is using the UI thread we are explicit about it and careful about it. When I did the Roslyn part of the project system/language service refactoring recently, there are totally places where I was on a background thread and a SwitchToMainThread() would have made my life easier. But the entire reason we found ourselves rewriting this code was to detach it from the UI thread; I'm cheering the fact that while I've been doing that the shell has been making more services available off the UI thread, which just negates the need for us to use JTF, _and_ means the UI will be less blocked. That's the ultimate win in my book.

**Comment by AArnott** [comment]
> I'd say the don't chuckle; ... When I did the Roslyn part of the project system/language service refactoring recently, there are totally places where I was on a background thread and a SwitchToMainThread() would have made my life easier. But the entire reason we found ourselves rewriting this code was to detach it from the UI thread;... I'm cheering the fact that while I've been doing that the shell has been making more services available off the UI thread, which just negates the need for us to use JTF, and means the UI will be less blocked.

Most components don't have the enviable opportunity to rewrite themselves like Roslyn did from the old C# project system and compiler. We have to make it easy to go from UI thread locked to async and enable switching to and from the UI thread easily or almost no one would have ever gone async in VS since they have to be allowed to do so progressively.

I've always maintained that JTF is a bridge between sync and async worlds. As such, if it serves its purpose well, one day we may not need JTF any more since VS will be 100% async.

**Comment by CyrusNajmabadi** [comment]
> I'd say the don't chuckle; this to me is the truth and why I've loved that Roslyn didn't use JTF. 

Agreed.  it was very much a *good* thing (regardless of TPL/JTF/VS/whatever).  It's a valuable thing to have to be extremely cautious and judicious about this stuff.  We've all experienced the component that is technically async but still has a terrible UI experience due to misusing the UI thread.  By *both* explicitly not having easy ways to get to hte UI thread, **and** also having it be that we would generally deadlock if someone even tried this, it forced Roslyn to design effective solutions in these scenarios.

To me, the "code just calls willy-nilly into the UI" is effectively as bad as the "STA code pumps on while taking locks" crazyness.  At least it's always explicit when you're going to UI thread.  But being simple and easy is not actually a virtue for me as i very much never want people to do it haphazardly in roslyn.  Sometimes it's good that dangerous stuff is hard.  It makes people really have to think very carefully and judiciously about things.  

> I've always maintained that JTF is a bridge between sync and async worlds

it's worth pointing out that i don't think that's what Roslyn has really used it for.  It seems primarily to have been used to replace someone components we wrote that were TaskSheduler-based for working with hte UI thread.  We had a system that did work (and did behave with TPL semantics we expected).  But, i think we're always happy to remove code if it turns out to be unnecessary.  there was just an mismatch in expectations.  We though "SwitchToMainThreadAsync" was equivalent to "Schedule to run on UI thread, with the same cancellation token guarantees as what TPL offers".  We've clearly learned the latter part isn't the case :)

Note: i'm not unhappy with us using it, and i think it works well for this purpose.  It's just an important detail we have to add to our mental toolkit (and hopefully to our analyzer toolkit) for future safety.

**Comment by CyrusNajmabadi** [comment]
> Or did you come up with your own way to allow UI thread-bound task continuations to execute even while you were blocking the UI thread? Did you rely on RPC anywhere?

We just designed things to never do this.  If you were on the UI thread, it was only acceptable to ever block a completely computation-bound bg task that itself never had any sort of UI need.  And, normally, we've tried to make it so that the UI thread just doesn't try to even block those guys.  We've had to in a small number of cases because there is no other way.  But, if we have requirements on other systems we consider that to be a responsibility of the UI thread, and not of the BG work we've kicked off.  It's kept us sane, and has worked well for pretty much everything that's been thrown at us so far :)

**Comment by CyrusNajmabadi** [comment]
> If we do change the behavior, we need to consider this, and I'd like your input on it. What should happen if you're already on the main thread but your token is canceled when you call STMTA, assuming alwaysYield: false?

Is there something wrong with cancellation at that point?  I do assume that any time i await a call that is passed a real cancellation token that it, well, cancel :)    This could be an 'await' to just any method-that-takes-cancellation, and any method-that-takes-cancellation may throw on the very first line, and that would be fine.  So i'm not seeing how things change to STMTA.  

Is there a reason to expect that it would behave differently, even if you were on the UI thread and called into it?  Thanks!

**Comment by CyrusNajmabadi** [comment]
> With JTF, there's always just one way to do it. It's simple, it works, and if you consider a UI thread dependency to be evil, it's easy to notice that you have one and block its introduction in a PR review.

Good point.  I can very much respect the value and sanity that brings to many teams :)

**Comment by CyrusNajmabadi** [comment]
> and you managed to convince all your partners to never add thread affinity to their code that you called as well

Basically, as long as VS has been around, you could never change thread affinity (i mean, i suppose you could make a STA component MTA... not sure if anyone ever did though).  So, we tried to get free-threaded components from lots of teams.  And, if we coudn't we just accepted the UI thread as a necessarily evil and tried to design things so that regardless of that evil, we weren't making things worse.

**Comment by sharwell** [comment]
>  I do assume that any time i await a call that is passed a real cancellation token that it, well, cancel 

I never think if it this way. I expect cancellation to trigger the fastest path to the end of the method, which would be successful completion if the remaining work is cheaper than throwing an exception.

**Comment by CyrusNajmabadi** [comment]
I wasn't saying that it *will* cancel.  Merely that it can cancel.  i.e. if given:

```c#
Foo(cancellationToken);
```

If i pass that with a cancelled token, then my expectation is absolutely that this can cancel (at any point or any time), including (but obviously nto limited to) the first line.


Ergo, the same holds true for STMTA.  if i call it with a canceled token, then having it throw seems utterly reasonable and in line with what anyone might think was going to happen for any cancellation aware method.

> I expect cancellation to trigger the fastest path to the end of the method

Looking at roslyn, it looks like that's what... one in a 100 occurrences**?  I can't even see anything in the framework that would generally make me think this is reasonably common.  

--

** Seriously, in roslyn fast-past checks happen 8 times. Cancellation throwing happens 800.  I would say that's practically noise for the former case.

**Comment by jasonmalinowski** [comment]
> I expect cancellation to trigger the fastest path to the end of the method, which would be successful completion if the remaining work is cheaper than throwing an exception.

How expensive is throwing an exception? How many equivalent IL instructions is it? "Do the cheapest" is definitely a good argument, do we have any numbers to know what that actually is?

**Comment by jasonmalinowski** [comment]
> If we fix more subtle bugs than we introduce (by a healthy margin) by making a behavioral change, then I expect I'll come around to being willing to throw while still on the UI thread. I don't expect to come around to the idea of waiving an available UI thread to wait and throw later on a threadpool thread as that would create a "by design" perf delay that makes cancellation slower than successful completion. So far, the data Jason is collecting supports the behavioral change.

Fully agreed that jumping back to the thread pool to raise the cancellation is silly, so let's not do that. And I'm also 100% for just an analyzer if we determine it is too much of a risk to change the behavior. More than anything else, I'd love it if owners of the code using SwitchToMainThreadAsync that did cancellation could look at their code and see if there are in fact issues there or not. It looks fishy, but all it takes is one case where it's clearly a regression and well, so much for a behavior change.

> My own argument that STMTA should behave like other async methods is somewhat weakened when we consider that it is not a typical async method.

Agreed fully: this is as magic of a method as you get; if there was a method where usual rules wouldn't apply, it's here!

**Comment by CyrusNajmabadi** [comment]
> Um, what? Are you saying, @CyrusNajmabadi that you wouldn't be surprised if canceling an async method would make the method take longer to complete than if you hadn't canceled it?

I'm saying: i find it to be completely acceptable and normal for a canceled method to throw a cancellation exception.  

It's literally the norm.  It's how we ourselves write 99%+ of all our cancellation code.  We've got a decade of experience telling us that this is fine.  So, as i mentioned on your original question: I am 100% ok if STMTA throws if it gets a canceled cancellation token.

**Comment by CyrusNajmabadi** [comment]
> How expensive is throwing an exception? How many equivalent IL instructions is it? "Do the cheapest" is definitely a good argument, do we have any numbers to know what that actually is?

Given that this is how roslyn basically behaves *all the time*, and we're massively concurrent+cancelling, i'd say: this is not a concern that we care about.

The only time i've ever had an issue here was at one point when the debugger had a massive perf hit when an exception was thrown.  In Roslyn this was then painful as you could get hundreds of cancels a second as things like typing kicked off work only to have them cancel a short while later. 

However, that issue was at least 5+ years ago, and there hasn't been any problems around cancellation and perf.  Indeed, as Jason mentioned, cancelling early is *great* for perf because we then don't run any computation that is now unnecessary since we already said "we don't want to run anything if cancellation happens".

**Comment by AArnott** [comment]
For the record, I'm not concerned about the perf cost of throwing an exception. I'm concerned about the perf cost of waiving access to our current UI thread and waiting for an arbitrarily long period of time for a threadpool thread to come along so we can throw the exception on the threadpool instead of the main thread.
Throwing from a threadpool thread wouldn't be a breaking change. Throwing from the main thread would be. But throwing from the main thread makes more sense from a perf perspective.

**Comment by CyrusNajmabadi** [comment]
> Throwing from a threadpool thread wouldn't be a breaking change. Throwing from the main thread would be. But throwing from the main thread makes more sense from a perf perspective.

Why would throwing from the main thread be a breaking change?  Isn't htat what most code that follows stmta will do when it itself uses that cancellation token?  isn't that what roslyn is doing here?

Do we have an expectation that that would actually break people?

**Comment by CyrusNajmabadi** [comment]
> If it could, it did its job and should report successful completion

I uess i view it exactly like '.ContinueWith'.  .ContinueWith could be 'successfully' doing it's job by executing the continuation code on the requested thread if it had gotten to that point and was ready to go.  

To me, i don't really see why STMTA is any different.  Both are just requests to "get to this point, and then do followup work".  I just see 'cancellation' as indicating "and don't bother doing this if things were canceled, because i indicated that i don't care."

**Comment by weltkante** [comment]
As a user of JTF outside VS I don't have a strong opinion about whether to allow for adjusting the behavior of STMTA, but introducing subtile semantic differences is something which should be weighted carefully against the utility in Roslyn. Maybe its better to have a dedicated (Roslyn specific) helper method for switching to the UI thread, which provides the desired semantics and use analyzers to control the TPL-to-JTF transition?

> I guess i view it exactly like '.ContinueWith'

I don't think this mindset is scalable to async Tasks vs. TPL tasks, or are you assuming this just for STMTA?

For async Tasks in general I don't see any way this point of view can be correct, the Tasks generated by the compiler for async methods don't know anything about cancellation tokens you passed in, and nobody forces async method implementations to check their tokens on each returning branch, so cancellation is inherently lazy by design and you shouldn't assume anything of that kind. If you care about cancellation in async Tasks you need to check it yourself if you want to write robust code.

**Comment by CyrusNajmabadi** [comment]
> or are you assuming this just for STMTA?

The question was about how STMTA should behve.  I was saying that i view it similarly to .ContinueWith, and i would prefer cancellation behave as similarly as possible.

**Comment by CyrusNajmabadi** [comment]
> so cancellation is inherently lazy by design and you shouldn't assume anything of that kind. 

AFAIK, it's not an assumption.  We discussed hte behavior of TPL *in depth* with the TPL authors as it was being created.  We were heavy consumers and were communicating with them regularly about what semantics we could expect and what was considered undefined behavior.  From all my recollections, this very much was by design.

> If you care about cancellation in async Tasks you need to check it yourself if you want to write robust code.

Why? This is like saying that to be robust, i also should not assume that TPL sets up the right fences across tasks, and one should do that themselves to ensure that a descendent task sees the writes of an antecedent task.

In reality this is unnecessary, and we've at least had verbal confirmation that that is the case (though trying to find official documentation for this sort of thing is painful).  So, should we then go update all our code accordingly because we don't trust these sorts of behaviors that have been baked into TPL for ages?

**Comment by weltkante** [comment]
You seem to have misread my comment. I was talking about cancellation of async Tasks (e.g. those generated by the compiler when writing async methods), not challenging the behavior of TPL. It's unfortunate that both use the same Task class.

Changing semantics for STMTA doesn't help reviewing any other async Task usage, cancellation is still lazy and the callers should check their tokens if early cancellation is important for them instead of lazy cancellation. As far as I'm concerned I'm reviewing STMTA cancellation like every other awaitable method. If you are interacting with async methods from external APIs you can't guarantee that they have checked their cancellation tokens on every returning branch, and even in Roslyn I assume going forward with the new completion APIs you have to consume async methods on interfaces implemented somewhere else in VS or even by VS extensions you can't control.

I understand that with TPL you didn't have to be explicit about cancellation checks, but if you transition away I'm arguing that you can't keep that property forever because it was dropped with the introduction of async Tasks alongside TPL tasks. So instead of special casing certain methods IMHO it scales better to just be explicit about cancellation everywhere and make the async/await assumption of lazy cancellation.

(And yes I acknowledge that for transitioning away from TPL you may want to bridge the semantic gap with things like STMTA but it may be a bit dangerous to trade one subtle semantic problem in Roslyn with adding a subtle semantic change for every consumer of the API. I don't know how likely it is to accidentally depend on not throwing when already being on the main thread, but it might be better to not require other consumers in VS to have to think about that edge case where they didn't have to before.)

Thats just my outside view though, as said I don't have a strong opinion and could deal with the change as far as our own codebase is concerned, just being a bit wary about the subtle semantic change against previous behavior.

**Comment by CyrusNajmabadi** [comment]
> You seem to have misread my comment. I was talking about cancellation of async Tasks (e.g. those generated by the compiler when writing async methods),

I guess i'm not sure what you mean by that then.  The compiler doesn't create any sort of special tasks.  It just uses whatever Task type you return (normally System.Threading.Task or ValueTask).  This is normally created with AsyncTaskMethodBuilder, with a bunch of callbacks indicating to which state things should transition as intermediary tasks complete.  But the operation of those tasks still operates as per how the TPL operates...

> Changing semantics for STMTA doesn't help reviewing any other async Task usage

The difference here, for me, is that the use of JTF here in roslyn was specifically is to supplant the usage of existing TPL mechanisms for moving to the UI threads. i.e. this isn't "any other async Task usage". This is specifically the "schedule the remainder of this work to happen over here".  

I do acknowledge that that means this method is more specialized in my mind.  And i do understand the argument that one can look at this and just think of this method as if it was just any random async-method.   I do get that argument.  I just feel that this is special enough to behave differently.

> I don't know how likely it is to accidentally depend on not throwing when already being on the main thread

To me, this is the important question.  I have a feeling that it would be pretty unlikely to hit anything problematic here.  Anything using cancellation here for the switching seems like it will just continue using that cancellation after that point, which means cancellation coudl happen on that thread without any problem.

But i do also acknowledge this could be considered a change in behavior that might have negative results for some clients as well.

**Comment by weltkante** [comment]
Clarifying my point in case you care about what I meant, I agree with the rest of your response.

> This is normally created with AsyncTaskMethodBuilder, with a bunch of callbacks indicating to which state things should transition as intermediary tasks complete. But the operation of those tasks still operates as per how the TPL operates

But that task created through AsyncTaskMethodBuilder (which I'm calling async Task) doesn't check any cancellation tokens, unless you do it explicitly in the method body. What you probably mean is that when you ContinueWith on an async Task that the returned task operates under TPL, but thats of course because ContinueWith produces a new Task according to TPL rules. You have to pass the cancellation tokens *again* to ContinueWith, TPL will not magically respect tokens passed to the original async Task, either.

So when you say you expected STMTA via await/OnCompleted to behave like ContinueWith I say thats a dangerous/unfounded assumption to take because any other async Task/Awaitable someone writes will by default *not* have that behavior either. (That doesn't argue against the rest of the discussion about making STMTA a special case where your assumption works.)

So if you have an async API consumed by Roslyn but implemented by VS or even an extension you can't just assume the Tasks returned by that API will respect your cancellation tokens the way ContinueWith does, you always have to check the cancellation tokens again (or forcefully switch to TPL by calling ContinueWith and pass the tokens again for implicit checking). Either way you need a second token check to have eager cancellation instead of lazy cancellation.

**Comment by AArnott** [comment]
Thanks, @weltkante. What I get from your argument is:

1. In general, *async* methods that accept `CancellationToken` are generally free to honor or not honor that token. They certainly *may* not check them as the last thing they do before each of their exit paths. And therefore, it's not a good idea to *rely* on async methods to check the token on your behalf if you rely on the token being honored.
1. It *is* a good idea to explicitly check the token yourself if you care about it.
1. We should not take subtle behavior changes in a platform assembly lightly, because it's hard to anticipate its full impact on existing users and we don't want to cause app failures nor violate customer trust.

And from Cyrus's response:

1. The STMTA method is *not* your typical async method, but because of its very special purpose and because it's one particular one that we *can* control. Therefore it is possible, justifiable, and convenient to *ensure* that it honors the token. This brings it inline with guarantees that TPL makes for *scheduled* tasks. This is useful for someone migrating from TPL scheduled tasks, and also probably everyone who calls STMTA. The risk of this subtle change is low because these callers are presumably prepared to deal with cancellation, even on the main thread.

I think all your asserted points are good ones.

FWIW, we haven't yet shipped a 16.0 versioned library, but we're about to. Making a behavioral change such as the one Cyrus is proposing would best be done at a major version increment so it aligns with customer expectations (per semver rules) that major version increments may include breaking changes. 
Our window for making changes to 16.0 is short.

At this point, I'm leaning toward the behavioral change, given the evidence @jasonmalinowski has found so far, and the point @CyrusNajmabadi has made that customers want to cancel if the token they passed in is canceled.

I'll discuss with the folks I'd have to convince to share the risk of regressions to see how they feel and will report back here. But further comments are of course welcome.

**Comment by CyrusNajmabadi** [comment]
> What you probably mean is that when you ContinueWith on an async Task that the returned task operates under TPL, but thats of course because ContinueWith produces a new Task according to TPL rules

Yes.  I thought that was clear from what i was saying, but if not i apologize.  It's been a long thread, and there's been a lot stated.  But i tried to make it clear that i felt STMTA was morally close enough to ContinueWith (since that is how you *today* run continuations on another scheduler) that it should behave the same way.  

I def acknowledge that there's no statement anywhere that that needs to be so.  And i'm def aware that changing to have that behavior is a subtle change and that i'm not in a position to make claims about.  

It's just what i would have preferred :)

**Comment by CyrusNajmabadi** [comment]
> At this point, I'm leaning toward the behavioral change, given the evidence @jasonmalinowski has found so far, and the point @CyrusNajmabadi has made that customers want to cancel if the token they passed in is canceled.

Another point of note: when we pointed out a subtle and potentially unintuitive part of TPL 5+ years ago, they attempted to go this path.  They felt the change would be more intuitive as well.  But, as it turned out, it led to breaks down the line.  So the change was reverted and the new behavior was brought in through an option.

So i'm *extremely* sympathetic to the concern that this could break people.  So much so that i have zero problem with that potential problem being a deal-breaker.  So, please, only make the change if you truly think it is the right thing to do.  If you think it isn't (possibly because it's too late now and we have to live with it), then i think it's 100% fine for Roslyn to be the one that has to change here.  Not my personal pref.  But this is not our library, and you're in the best position to make this call for your entire userbase.

Thanks!

**Comment by lifengl** [comment]
I saw the proposal in the email thread.  I don't think that I agree with the second half of the proposal.  If the cancellationToken is triggered in the middle of the transition, and the continuation reaches the UI thread first, I don't think it should throw the CancellationException on the UI thread.  Basically, it is always a race condition when the cancellationToken is triggered in the middle of a task, and there is no reason to ensure it will be honored.  We can consider it is similar to the case that the task has already been completed before the CancellationToken is triggered, but the task continuation hasn't been executed. (Even after the change, the CancellationToken can always triggered right after we check it on the UI thread, and before it executes the continuation code.)  It sounds strange, and logically incorrect to me that it will throw the Cancellation exception on the UI thread at that point.  Just like Andrew pointed out, it would run the exception logic (and rest logic) in the UI thread (instead of the original context), which is unexpected.

**Comment by CyrusNajmabadi** [comment]
> Basically, it is always a race condition when the cancellationToken is triggered in the middle of a task, and there is no reason to ensure it will be honored

There is no race condition.  The cancelling has hapened on the thread that we are transitioning to.  Since two tasks cannot execute on the same thread at the same time, then there is a strict ordering here of when things can happen.  Either the cancel happened before, and thus we can read and safely cancel when we transition to that thread.  Or, it happens afterwards... in which case, this is moot.  

> Just like Andrew pointed out, it would run the exception logic (and rest logic) in the UI thread (instead of the original context), which is unexpected.

Why is that unexpected?  It's certainly how TPL works.  Indeed, as mentioned, you had to always assume such a thing *was* possible since there is no guarantee of when a called async method might cancel.  i.e. it *might* cancel as the very last thing it does.  So, in that case, this would happen on that thread anyways.

**Comment by CyrusNajmabadi** [comment]
> But on the other hand, in the normal await pattern, await MyAsyncMethod(), will always throw the exception in the original context. It never throws and runs the continuation in a random context.

It hardly seems random here :)    As stated a few times (and well put by Andrew here https://github.com/Microsoft/vs-threading/issues/434#issuecomment-448647136), this is particularly about STMTA and the semantics *specifically* as a thread-swapping mechanism, much more akin to `.ContinueWith(.. Main-Thread-Scheduler..)`.

**Comment by lifengl** [comment]
It may or may not impact some of try/retry to switch to the UI thread logic.  Some of our logic is done in the pattern like, SwitchToUIThread with a time-out (using a cancellationToken). If it works, we finish the work on the UI thread. If it doesn't get into UI thread on time, we run a different set of logic.  That includes the logic to scan task dependencies to verify whether we are in a critical task chain (like blocking the UI), if that is true, we will try to switch without a time-out.  If it is not true, we will delay processing it, and try it later.  The reason we didn't do the dependencies check at the beginning is that it is an overhead, and might be slow.  The change would mean that this logic might run in the UI thread, when the time-out happens to be triggered.  Of course, this can easily be updated.  But my point is that throwing on the UI thread may impact some existing logic.

**Comment by CyrusNajmabadi** [comment]
> But my point is that throwing on the UI thread may impact some existing logic.

Understood.  Thanks for the example.

**Comment by AArnott** [comment]
Given 16.0 is closing down and there wasn't enough appetite among other stake holders to take the risk of changing this behavior, I'm closing this as Won't Fix. This doesn't mean it wasn't a good idea and perhaps even the right choice, but it just means our window for making such a behaviorally significant change was a major version increment and as 16.0 is behind us now, that leaves #435 to mitigate this.

**Comment by AArnott** [comment]
More evidence from other VS code suggests that always throwing when the token is canceled may be the right policy. I'm preparing a change to do this.
